### ISSUE TITLE
Feature Request: Call "Say" from modal True screens...

### ISSUE THREAD

**Issue by RedLightning42** [renpy/renpy#3594]
Working on a game where I was asked to make inventory descriptions display as if the characters were explaining them. This includes potentially multiple lines of dialogue and other code (to add new inventory items, etc.). 

For now, Fen on the Discord has written a small script to call "Say" into a new context above the inventory, and return when the label I call ends, but it would be nice to have an official mechanic to display a single line or even full label's worth of dialogue OVER screens that are "modal True".

(sorry if the format of this request is wrong, I'm still new to GitHub)...

**Comment by renpytom** [comment]
I like the idea of this, but I'm not sure this can be implemented in a way we support.

**Comment by RedLightning42** [comment]
See, then either I'm missing something, or my game wouldn't work without Fen's code. Even from a "modal False" (which I assume is default) I'm having trouble making dialogue appear over the screen then return to the screen. Here's what Fen whipped up as an example, and it makes my game much easier to write...

`init python: #Borrowed code to show dialogue OVER other screens
    class SayOnScreen(Action):

        def __init__(self, label, *args, **kwargs):
            self.old_clear_layers = config.context_clear_layers
            self.label = label
            self.args = args
            self.kwargs = kwargs

        def __call__(self):
            renpy.config.context_clear_layers = [ ]
            renpy.call_in_new_context('intermediary_sayonscreen',
                self.label, self.old_clear_layers,
                *self.args, **self.kwargs)

label intermediary_sayonscreen(lbl, original_layers, *args, **kwargs): #Borrowed code to show dialogue OVER other screens
    $ config.context_clear_layers = original_layers
    show screen absorb_input()
    $ renpy.call(lbl, *args, **kwargs)
    hide screen absorb_input
    return

screen absorb_input(): #Borrowed code to block anything below while showing dialogue OVER other screens
    layer "above_screens" #added to bump this to a higher layer so I can blur the background.
    zorder 1000
    key 'dismiss' action Function(renpy.ui.saybehavior, allow_dismiss=renpy.config.say_allow_dismiss)
    button:
        xysize (config.screen_width, config.screen_height)
        action Function(renpy.ui.saybehavior, allow_dismiss=renpy.config.say_allow_dismiss)
        `

**Comment by Gouvernathor** [comment]
I'm not sure I understand, is the purpose here to make the screen be insensitive while the say screen is displayed above it, and then make it sensitive again ? Or is it the opposite, to have the clicks only get to the screen without advancing and dismissing the dialog ?

If it's the former, I think the simplest solution would be to use the `sensitive` clause of the screen. Does that interfere with `modal True`, is that the problem ?
If it's the latter, how is the dialog supposed to be dismissed ?

**Comment by RedLightning42** [comment]
Honestly, I'm not sure , but the issue I was having is that my Inventory is "modal True" and I want to display multiple lines of dialogue OVER that , which currently means a label, so I can run dialogue and even python commands, then return to the Inventory. Not my code, I just implimented it.

**Comment by Gouvernathor** [comment]
Ok, I think I got it, so your goal is the first thing I listed, and the problem is that the modalness of the screen prevents clicks from reaching the say screen and make it go away.

**Comment by Andykl** [comment]
If I understand the problem correct, I believe all you need is already possible in RenPy. It may look like:
```py
screen inventory():
    modal True
    inv_item action Return(item_label)
    textbutton "Return" action Return("exit")

label call_inventory():
    label .repeat:
    call screen inventory
    if _return == "exit":
        return

    show screen inventory(_zorder=-1)  # show inventory again
    call expression _return
    jump .repeat
```

**Comment by shawna-p** [comment]
Hello, as the person who provided the original SayOnScreen code I feel I should add my two cents.
Unfortunately @Andykl that solution does not accept mouse input due to the screen being `modal True`. You'll get stuck after jumping to the label with the dialogue/item description. Even setting the screen to `sensitive False` will not solve this absorbed input problem.
Further, while this could be awkwardly circumvented by having a modal and non-modal version of a screen which is then `use`d by some parent screen, it doesn't solve the larger issue either - in some cases, a player may have multiple screens showing, and it would be inconvenient to impossible to re-show these screens (+ any passed information) just so you could display dialogue behind them.
This issue has come up 3 separate times in the past week on the Ren'Py server alone, which I think indicates it's an active problem where people are searching for a solution. The crux of it is that people would like a way to display dialogue over whatever in-game screens are currently showing. It's a common enough use case, for example, to have characters comment on a player's inventory items, or to have a hint system to nudge the player where to go on an in-game map.
I won't deny my solution is rather hacky - it essentially enters a new context so dialogue can be shown again despite being on a called/modal screen and then fudges player input to progress dialogue - but it does save the developer from having to juggle which screens are shown and re-called, or from creating a separate say screen and being unable to access Character property information to style it according to the speaker. If there was a better way of going about it, I imagine there are several games which could take advantage of it.

**Comment by RedLightning42** [comment]
As mentioned above, I encountered MULTIPLE issues trying to make it look like my characters were talking about things in the inventory (replacing tooltips with dialogue). And aside from all that, I recently moved my navigation from a menu to a "modal False" screen, but EVEN THEN I can't seem to figure out how to get "Say" to appear OVER the "map" screen with multiple lines, then return to the map... If I use normal Call or Jump, it hides the screen and shows black (unless there's a new Scene in the label I'm going to) and I want something that will show up ON the screen.

Think Point and Click games, where you can click on an object you see (imagebutton or hotspot imagemap). Sometimes it takes you to a new location (easy enough) but other times the MC will describe what they're seeing or have a "REASON" they can't go somewhere or do something... It all happens in one screen, but in Ren'PY we have to mix a "map" screen with "Say" and it just doesn't work right now without Shanwna-p's (Fen's) code... Their code also let me bump "Say" effectively to a new layer so I could separate it from the Screen Layer, blur Screen and Master, and display Inventory AND dialogue over the blur.

I realize that some/all of this is not intended in Ren'PY, but I chose this engine/language because I figured with Python, it would be the most flexible for making a less-traditional VN Game. That said, take a look at things that exist, like Danganrompa or Zero Escape, both of them (AFAIK) have the ability to navigate visually and then have conversations in the same view... I'm now learning this is very difficult in Ren'PY without the "new context hack"... It's just something that the community would like you to consider for a future release... Thanks!

**Comment by RedLightning42** [comment]
Changed the title after my recent learning about screens with modal defaulted to False. As in when I built screens full of imagebuttons, I didn't set Modal to anything (so it take the default) and I still have trouble...

**Comment by RedLightning42** [comment]
if I show a non-modal screen, it advances dialogue if I click anywhere that is not an imagebutton. so we're back to the same problem or making EVERY screen modal True, and then yes, same problem.

**Comment by mal** [comment]
Building atop Andy's code, we can get the most of the desired functionality without too much magic:
- Normal dismiss behaviour to advance dialogue - space, enter, mouse click, etc.
- No re-showing different versions of the screen or screens.
- Blurring screens while the explanation plays.

The main concepts are that the sensitive and modal statements allow expressions, so we can change their states based on external factors (in this instance whether the wait_for_user screen is showing; and that calling an empty screen will receive the first (inter)action taken on any screen, allowing waiting for a return value without hiding the inventory (i.e. the difference between show and call screen noted by Gouvernathor).

```rpy
init python:
    renpy.add_layer("middle", above="master")

screen inventory():
    layer "middle"
    add Solid("#244")
    modal renpy.get_screen("wait_for_user") # if you don't need modal, omit this line
    sensitive renpy.get_screen("wait_for_user")
    textbutton "Item" action Return("item_label") align (.5, .5)
    textbutton "Return" action Return("exit") align (.5, .6)

screen wait_for_user():
    pass

label call_inventory():
    label .repeat:
    show screen inventory()
    call screen wait_for_user()
    if _return == "exit":
        hide screen inventory
        return

    show layer middle at unfocus
    call expression _return
    show layer middle at reset
    jump .repeat

label item_label:
    "Hello, I'm an item description!"
    return

transform unfocus:
    blur 10
```

It won't implicitly solve the lots of stacked screens part, but by placing any such screens on the new "middle" layer, and using some variation of modal and sensitive statements in them should make it possible to get it functioning as desired.

**Comment by Booplicate** [comment]
> The main concepts are that the sensitive and modal statements allow expressions

Correct me if I'm wrong, but `modal` doesn't take an expression, it's being evaluated only once during screen definition.

**Comment by renpytom** [comment]
Modal is evaluated at the moment the screen is shown, but I could change that without much problem, I think.

**Comment by RedLightning42** [comment]
```
init python:
    renpy.add_layer("middle", above="master")

screen inventory():
    layer "middle"
    add Solid("#244")
    modal renpy.get_screen("wait_for_user") # if you don't need modal, omit this line
    sensitive renpy.get_screen("wait_for_user")
    textbutton "Test" action Return("testing") align (.5, .4)
    textbutton "Item" action Return("item_label") align (.5, .5)
    textbutton "Return" action Return("exit") align (.5, .6)

screen wait_for_user():
    modal True

label call_inventory():
    label .repeat:
    show screen inventory()
    call screen wait_for_user()
    if _return == "exit":
        hide screen inventory
        return

    show layer middle at unfocus
    call expression _return
    show layer middle at reset
    jump .repeat

label item_label:
    "Hello, I'm an item description!"
    return

transform unfocus:
    blur 10

label start:
    scene bg leaving
    "this is a test"
    show xyla at left
    e "with a sprite"
    show screen inventory
    "all done"
    "exiting now"
    return

label testing:
    show xyla at left
    e "does this work?"
    hide xyla
    "maybe"
    return
```

made my own demo... I'm not getting something, because the moment I show the inventory, the dialogue just keeps going till the game ends...

**Comment by mal** [comment]
You're not calling the label that's managing the inventory screen:
```diff
 label start:
     scene bg leaving
     "this is a test"
     show xyla at left
     e "with a sprite"
-    show screen inventory
+    call call_inventory
     "all done"
     "exiting now"
     return
```

By default `xyla` will show on the master layer, behind the inventory screen which is on the middle layer. If you want her in front you'd need to the sprite on a layer above that (unless the default layer for that tag is already higher).

**Comment by RedLightning42** [comment]
So I fixed SOME of my issues by putting the modal False screens on the Master Layer since they're meant to be a background... Great, one down. Still would like the ability to call a label over a modal True screen, even if that means setting it temporarily to sensitive false or something. But to be able to put dialogue (and possibly characters) from a label OVER my Inventory is the whole point of the code we started with... back to this another day when my head's not exploding...

**Comment by RedLightning42** [comment]
One more update for today. I got it to work, BUT if you have a button to call the inventory (like I do) and you click it while dialogue is going on, that dialogue stays on screen and takes precedence... It's not exactly what I was hoping YET...

**Comment by RedLightning42** [comment]
Look, I'm going to bring this up again cause I just saw someone in the discord trying to duplicate Say functionality AGAIN... I'm still having all kinds of issues mixing screens and Say. I'm CLOSE, but there's visual things using Fen's code for a new context...

What I have:
All BG files are on the layer Back (under Master) so I can blur them separate from sprites if needed and any imagemap Nav screens are on the same layer. Sprites and Say/Choice are still on Master.
INVENTORY is on a new "Above_Screens" layer for a similar reason (can background blur everything behind it, including dialogue or choices currently showing (replaces the old transparent grey method). 
SayOnScreen handles the dialogue calls (one line, or sometimes entire labels with 2-3 characters and python calls) from the inventory.
The nav screens use a mix of normal Jumps and SayOnScreen to show dialogue over the BACKGROUND, but it HIDES THE SCREEN so any imagebuttons vanish. Sometimes useful sometimes not... (why the heck can't we leave a screen visible on a jump or call?) I'd prefer to be able to leave the buttons for some cases (and pick them myself).
I also learned that I don't understand how to show a screen as a background, but have the buttons be non-functional till a certain point, had to make a separate version...

PROBLEMS with this method ALSO include:
SayOnScreen (new context use of Say) often leaves a textbox frame (sheet of paper image) to flicker for a frame when I jump between screens for navigation, even though no textbox is showing when I switch. 
For some reason SayOnScreen doesn't respect that when text is slow typing with CPS, the first click should display the rest of the line and the second click should advance, it just advances on the first click...

All this to say, it would be NICE to be able to recreate (or have two [multiple?]) versions of the Say Screen to display for different reasons. This all started because I can be in the middle of dialogue, pull up my Inventory (modal True) AND want to use Say inside the inventory to describe items or have conversations about them...

**Comment by GetOutOfMyLab** [comment]
I figured I'd add my $0.02, since my problem was what initially spawned Fen's SayOnScreen. The use case was having a computer that would open a website or app that is pretty much a popular social media photo sharing site. While viewing some photos, sometimes the player character has a few lines of dialogue as they're thinking about what's going on in the photo. Interaction is as follows: Open the website (called screen) -> Click a user's profile (shown screen) -> Click a photo from the profile (shown screen). On the shown screen is where I needed dialogue sometimes. 

Shortly after Fen helped me with a work around for this, several other people suddenly came into the discord with similar needs. Think of a detective story while examining clues and the player character and their assistant, for example, are discussing the clue. Of course, there is the classic inventory example where there may be a few lines of descriptive dialogue while examining inventory items. 

There's definitely a desire for people to want to be able to do such things, so I hope it could somehow be officially supported at some point.

**Comment by badpunman** [comment]
Yeah, this is a bit of a pain currently to figure out. I improvised from the above a sort of solution that somewhat works with some caveats, but i'm not sure how well this scales to more use cases or the practicality of it. 

```
init python:
    renpy.add_layer("lowest", below="master")

define chatOverInventory = False

screen wait_for_user():
    null 

transform unfocus:
    blur 10

# 'Call' a screen. 
#    If it returns tuple of ('chat', 'label') the label is called and chat/master layer is rendered above the screen.
#    When the label returns, screen's return value is re-awaited.
#    If it returns tuple of ('exit', None) this call exits.
# Can't stack many call_screenCustom calls.
# Screen must be in lowest layer. Screen must have "sensitive chatOverInventory == False"
# The scene/master-layer should be cleared before calling this, to ensure nothing is rendered on top of the screen
# Screen MUST return tuple values
label call_screenCustom(screenname): 
    label .repeat:    
    show screen expression screenname
    call screen wait_for_user() 

    if _return[0] == "exit":
        hide screen expression screenname
        return
    
    if _return[0] == "chat":
        $ chatOverInventory = True    
        show layer middle at unfocus
        call expression _return[1] from fromcall_screenCustom
        show layer middle at reset        
        $ chatOverInventory = False
    
    jump .repeat
```

```
image Eileen = "Eileen.png"
define e = Character("Eileen")

screen inventory():
    layer "lowest"    
    add Solid("#423655")        
    modal False  
    sensitive chatOverInventory == False
    textbutton "Test" action Return(("chat", "testing")) align (.5, .4)
    textbutton "Item" action Return(("chat", "item_label")) align (.5, .5)
    textbutton "Return" action Return(("exit", None)) align (.5, .6)

label item_label:
    show Eileen at left
    with move
    e "Hello, I'm an item description!"
    return

label testing:        
    show Eileen at  right
    with dissolve
    e "Huh."    
    return

label start:
    scene bg sky                
    hide bg
    call call_screenCustom("inventory")
    scene bg sky    
    "exiting now"
    return
```
